### PR TITLE
Add support for retrieving listening address of websocket Server

### PR DESCRIPTION
**Pull request by michaelbeaumont**
It's still a little cumbersome to use because `Start()` blocks but this makes it possible to use `0` as the port and let the system pick a port then later see which port was chosen, useful for testing.

### CODE REVIEW

**Comment by lorenzodonini** [comment]
I like the idea of splitting the listen and serve up.

I recently added some "hacky" tests on the errors channel, for increasing coverage. I'll quickly investigate where exactly they are breaking and suggest a solution to the failing test.

**Comment by lorenzodonini** [comment]
> It's still a little cumbersome to use because `Start()` blocks but this makes it possible to use `0` as the port and let the system pick a port then later see which port was chosen, useful for testing.

You could also modify the `Start` to accept the full listen address (i.e. `:8080`). That way, you're not just limited to the port. Just an idea.

**Comment by michaelbeaumont** [comment]
Yeah, I was going to add that separately but changing the listening IP is useful too. But unrelated to setting port to `0`, right?

File: ws/websocket.go [hunk 1]
```diff
@@ -316,17 +325,28 @@ func (server *Server) Start(port int, listenPath string) {
...
 		server.httpServer = &http.Server{}
 	}
+
 	addr := fmt.Sprintf(":%v", port)
 	server.httpServer.Addr = addr
 	server.httpServer.Handler = router
+
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		server.error(fmt.Errorf("failed to listen: %w", err))
```
Comment (lorenzodonini) [inline ws/websocket.go:335]: Needs a return after throwing an error, otherwise it will panic on the `ln.Addr()`, since `ln` is nil.

**Comment by lorenzodonini** [comment]
> Yeah, I was going to add that separately but changing the listening IP is useful too. But unrelated to setting port to `0`, right?

Yes, it's unrelated. Feel free to add it in a follow-up MR. It would be a breaking change as well, so no rush.

**Review by lorenzodonini — APPROVED** [review]

**Pull request merged**